contracts/exchange/OneSidedExchange.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+pragma abicoder v2;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+struct Price {
+    uint256 price;
+    uint256 interval;
+}
+
+struct PriceThreshold {
+    uint256 maxSellAmount;
+    uint256 maxBuyAmount;
+    uint256 interval;
+}
+
+/**
+ * @title One Sided Exchange
+ *
+ * The contract which represents a one sided exchange
+ at desired prices.
+ */
+contract OneSidedExchange is ReentrancyGuard, Ownable {
+    event Deposit(address token, uint256 amount);
+    event Withdraw(address token, uint256 amount);
+    event SwapSuccess(
+        address indexed trader,
+        address tokenA,
+        address tokenB,
+        uint256 tokenAAmount,
+        uint256 tokenBAmount
+    );
+    error InvalidAmount(string message);
+    error NoPriceExists(string message);
+
+    mapping(address => Price) internal _buyPrices;
+    mapping(address => Price) internal _sellPrices;
+    mapping(address => PriceThreshold) internal _thresholds;
+    mapping(address => uint256) internal _buyAmounts;
+    mapping(address => uint256) internal _sellAmounts;
+
+    constructor() Ownable(msg.sender) {}
+
+    function withdraw(address token, uint256 amount) public nonReentrant onlyOwner {
+        if (amount <= 0) {
+            revert InvalidAmount("Invalid amount");
+        }
+
+        if (ERC20(token).balanceOf(address(this)) < amount) {
+            revert InvalidAmount("Not enought tokens balance to withdraw");
+        }
+
+        _withdraw(token, amount);
+    }
+
+    function deposit(address token, uint256 amount) public nonReentrant onlyOwner {
+        if (amount <= 0) {
+            revert InvalidAmount("Invalid amount");
```

Comment:
Often error parameters has some incorrect variable that FE shows user, not just string that's copy of error title
so here I'd like to have InvalidAmount(amount)